io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -715,77 +703,43 @@ impl LvsLvol for Lvol {
-            let mut rsp = nvmf_req.response();
-            let nvme_status = rsp.status();
-
-            nvme_status.set_sct(0); // SPDK_NVME_SCT_GENERIC
-            nvme_status.set_sc(match errno {
-                0 => 0,
-                _ => {
-                    error!("vbdev_lvol_create_snapshot errno {}", errno);
-                    0x06 // SPDK_NVME_SC_INTERNAL_DEVICE_ERROR
-                }
-            });
-
-            // From nvmf_bdev_ctrlr_complete_cmd
-            unsafe {
-                spdk_nvmf_request_complete(nvmf_req.0.as_ptr());
-            }
-        }
-
-        let c_snapshot_name = snapshot_name.into_cstring();
-        unsafe {
-            vbdev_lvol_create_snapshot(
-                self.as_inner_ptr(),
-                c_snapshot_name.as_ptr(),
-                Some(snapshot_done_cb),
-                nvmf_req.0.as_ptr().cast(),
-            )
-        };
-
-        info!("{:?}: creating snapshot '{}'", self, snapshot_name);
-    }
-
-    /// Create snapshot for local replica
-    async fn create_snapshot_local(
+#[async_trait(?Send)]
+impl SnapshotOps for Lvol {
+    type Error = Error;
+    /// Create Snapshot Common API for Remote Device.
+    async fn create_snapshot(
         &self,
-        io: *mut spdk_bdev_io,
-        snapshot_name: &str,
-    ) {
-        extern "C" fn snapshot_done_cb(
-            bio_ptr: *mut c_void,
+        snap_param: SnapshotParams,
+    ) -> Result<(), Error> {
+        extern "C" fn snapshot_create_done_cb(
+            arg: *mut c_void,
             _lvol_ptr: *mut spdk_lvol,
             errno: i32,
         ) {
+            let s = unsafe { Box::from_raw(arg as *mut oneshot::Sender<i32>) };
             if errno != 0 {
-                error!("vbdev_lvol_create_snapshot errno {}", errno);
+                error!("vbdev_lvol_create_snapshot failed errno {}", errno);
             }
-            // Must complete IO on thread IO was submitted from
-            Mthread::from_ptr(unsafe {
-                spdk_bdev_io_get_thread(bio_ptr.cast())
-            })
-            .with(|| Nexus::io_completion_local(errno == 0, bio_ptr));
+            s.send(errno).ok();
         }
 
-        let c_snapshot_name = snapshot_name.into_cstring();
+        let c_snapshot_name = snap_param.name().unwrap().into_cstring();
+        let (s, r) = oneshot::channel::<i32>();
         unsafe {
             vbdev_lvol_create_snapshot(
                 self.as_inner_ptr(),
                 c_snapshot_name.as_ptr(),
-                Some(snapshot_done_cb),
-                io.cast(),
+                Some(snapshot_create_done_cb),
+                cb_arg(s),
             )
         };
-
-        info!("{:?}: creating snapshot '{}'", self, snapshot_name);
+        r.await
+            .expect("snapshot_create_done_cb")
+            .to_result(|error| Error::SnapshotCreate {
+                source: Errno::from_i32(error),
+                msg: format!("Snapshot Created with ResultCode {error}"),
```

Comment:
IMHO better to remove msg from here and add it to the snafu display as that's the whole point of using such as library.